### PR TITLE
Multi join prep - `sql_join_vars()`

### PR DESCRIPTION
**Pull request by mgirlich**
Prepares the multi join queries #984.

### CODE REVIEW

**Comment by hadley** [comment]
Can you please remind me of the context for this PR?

**Comment by mgirlich** [comment]
The multi join PR #984 is a bit bigger. I thought it would make it easier for you to review it if I split off some parts before.
This PR refactors the logic that generates the `SELECT` clause in a JOIN query.

**Review by hadley — COMMENTED** [review]
Sorry, I actually wrote these ages ago but somehow forgot to submit 😬

File: R/db-sql.R [hunk 1]
```diff
@@ -387,6 +387,94 @@ sql_join.DBIConnection <- function(con, x, y, vars, type = "inner", by = NULL, n
   )
 }
 
+sql_multi_join_vars <- function(con, vars, table_names, all_vars_list) {
```
[the remark this answers is not in the document]
Comment (mgirlich) [reply R/db-sql.R]: I added some examples and some tests. This should hopefully give you a feeling for the data structure.

File: tests/testthat/test-query-join.R [hunk 2]
```diff
@@ -37,3 +37,41 @@ test_that("sql_on query doesn't change unexpectedly", {
   expect_snapshot(semi_join(lf1, lf2, sql_on = "LHS.y < RHS.z"))
   expect_snapshot(anti_join(lf1, lf2, sql_on = "LHS.y < RHS.z"))
 })
+
+test_that("sql_multi_join_vars works", {
```
Comment (hadley) [inline tests/testthat/test-query-join.R:41]:
```suggestion
test_that("sql_multi_join_vars generates expected SQL", {
```

File: tests/testthat/test-query-join.R [hunk 3]
```diff
@@ -37,3 +37,41 @@ test_that("sql_on query doesn't change unexpectedly", {
...
+        var = list("x", "a", "b"),
+        table = list(1L, 1L, 2L)
+      ),
+      table_names = c("left", "right"),
+      all_vars_list = list(c("x", "a"), c("x", "b"))
+    ),
+    sql("`left`.*", b = "`b`")
+  )
+
+  expect_equal(
```
Comment (hadley) [inline tests/testthat/test-query-join.R:59]:
```suggestion
  # full_join(lf(x, a), lf(x, b), by = "x")
  expect_equal(
```

File: R/query-join.R [hunk 4]
```diff
@@ -44,100 +44,168 @@ sql_render.join_query <- function(query, con = NULL, ..., subquery = FALSE, lvl
 
 # SQL generation ----------------------------------------------------------
 
+#' @param vars tibble with three columns:
+#'   * `name` `<character>`: variable name in output.
+#'   * `table` `<list_of<integer>>`: tables index in database.
```
[the remark this answers is not in the document]
Comment (mgirlich) [reply R/query-join.R:49]: At the time the `vars` structure is created the table names are not yet known. They are only known later when it was made sure they are unique.

File: R/query-join.R [hunk 5]
```diff
@@ -44,100 +44,168 @@ sql_render.join_query <- function(query, con = NULL, ..., subquery = FALSE, lvl
...
+
+  # Handle `full_join()` which must COALESCE join cols and can never use *
+  if (any(ns > 1)) {
+    out <- purrr::map2(
+      vars$table, vars$var,
+      function(table_ids, vars) {
+        if (length(table_ids) > 1) {
+          sql_expr(
+            COALESCE(
+              !!sql_table_prefix(con, vars[[1]], table = ident(table_names[[1]])),
```
[the remark this answers is not in the document]
Comment (mgirlich) [reply R/query-join.R:110]: Theoretically, this shouldn't matter (`full_join()` is never combined with another join) but it still makes sense.

**Pull request merged**